File: app/billing.py
import calendar
import json
from dataclasses import dataclass
from datetime import date, timedelta

from app.db import Bill, Occupancy


def _overlap_days(a_start: date, a_end: date, b_start: date, b_end: date) -> int:
    start = max(a_start, b_start)
    end = min(a_end, b_end)
    if start > end:
        return 0
    return (end - start).days + 1


def unit_person_days(unit_occupancies: list[Occupancy], bill: Bill) -> int:
    total = 0
    for occ in unit_occupancies:
        days = _overlap_days(occ.start_date, occ.end_date, bill.start_date, bill.end_date)
        total += days * occ.tenant_count
    return total


@dataclass
class UnitShare:
    unit_id: int
    unit_name: str
    person_days: int
    amount: float


def split_bill(bill: Bill, unit_occupancies_map: dict[int, list[Occupancy]]) -> list[UnitShare]:
    """Split bill by person-days across its assigned units.

    unit_occupancies_map: unit_id -> list of Occupancy rows for that unit.
    """
    shares: list[UnitShare] = []
    person_days_by_unit: dict[int, int] = {}
    for assignment in bill.assignments:
        uid = assignment.unit_id
        occs = unit_occupancies_map.get(uid, [])
        person_days_by_unit[uid] = unit_person_days(occs, bill)

    total = sum(person_days_by_unit.values())
    for assignment in bill.assignments:
        uid = assignment.unit_id
        pd = person_days_by_unit[uid]
        amount = (pd / total * bill.amount) if total > 0 else 0.0
        shares.append(
            UnitShare(
                unit_id=uid,
                unit_name=assignment.unit.name,
                person_days=pd,
                amount=round(amount, 2),
            )
        )
    return shares


def bill_due_date(bill_end: date) -> date:
    """First day of the month following the bill's end date."""
    if bill_end.month == 12:
        return date(bill_end.year + 1, 1, 1)
    return date(bill_end.year, bill_end.month + 1, 1)


def month_bounds(year: int, month: int) -> tuple[date, date]:
    start = date(year, month, 1)
    if month == 12:
        next_month = date(year + 1, 1, 1)
    else:
        next_month = date(year, month + 1, 1)
    return start, next_month - timedelta(days=1)


def next_month(today: date) -> tuple[int, int]:
    if today.month == 12:
        return today.year + 1, 1
    return today.year, today.month + 1


def recurring_instances(rb, today: date) -> list[tuple[date, date]]:
    """Return (start, end) date pairs for all bill periods that should exist up to today.

    Deduplication key per recurrence type:
      daily   → (rb.id, that day)           — 1-day period
      weekly  → (rb.id, Monday of week)     — Mon–Sun period
      monthly → (rb.id, 1st of month)       — full-month period
      yearly  → (rb.id, 1st of that month)  — full-month period
    """
    start = rb.start_date
    recurrence = rb.recurrence
    config: list[int] = json.loads(rb.recurrence_config or "[]")
    instances: list[tuple[date, date]] = []

    # Optional cap: never generate periods that begin after the template's end_date.
    end_cap = getattr(rb, "end_date", None)
    if end_cap is not None and end_cap < today:
        today = end_cap

    if recurrence == "daily":
        d = start
        while d <= today:
            instances.append((d, d))
            d += timedelta(days=1)

    elif recurrence == "weekly":
        # config = weekday indices [0=Mon..6=Sun]; default Monday
        active_days = sorted(config) if config else [0]
        # Walk week by week (Monday-anchored); emit one bill per week when any
        # selected weekday falls on or after start_date and on or before today.
        monday = start - timedelta(days=start.weekday())
        while monday <= today:
            sunday = monday + timedelta(days=6)
            for wd in active_days:
                trigger = monday + timedelta(days=wd)
                if trigger >= start and trigger <= today:
                    instances.append((monday, sunday))
                    break
            monday += timedelta(days=7)

    elif recurrence == "monthly":
        # config = day-of-month list [1-31]; default 1st
        active_days = sorted(config) if config else [1]
        y, m = start.year, start.month
        while (y, m) <= (today.year, today.month):
            _, last = calendar.monthrange(y, m)
            for day in active_days:
                trigger = date(y, m, min(day, last))
                if trigger >= start and trigger <= today:
                    instances.append((date(y, m, 1), date(y, m, last)))
                    break
            m += 1
            if m > 12:
                y, m = y + 1, 1

    elif recurrence == "yearly":
        # config = month numbers [1-12]; default January
        active_months = sorted(config) if config else [1]
        for y in range(start.year, today.year + 1):
            for month in active_months:
                _, last = calendar.monthrange(y, month)
                bill_start = date(y, month, 1)
                if bill_start >= start and bill_start <= today:
                    instances.append((bill_start, date(y, month, last)))

    return instances

File: docs/js/billing.js
/**
 * Bill-splitting + recurring-bill generation.
 *
 * Splitting is **pro-rated against the full bill period**:
 *   • A unit's actual contribution = Σ over its occupancies of
 *     (tenant_count × inclusive overlap days with the bill period).
 *   • Each unit is charged at a fixed per-person-day rate =
 *     bill.amount / (full-occupancy person-days), where full-occupancy
 *     person-days = Σ over assigned units of (the unit's tenant_count ×
 *     the full bill-period length). So a tenant present for only half the
 *     period pays half; the vacant remainder is simply NOT billed (it is
 *     not redistributed to the other units).
 *   • When every assigned unit occupies the entire bill period, the
 *     reference equals the actual person-days, so the full amount is
 *     recovered and the split reduces to plain person-day proportions.
 *   • Per-unit amounts are rounded to 2 decimals.
 *   • A unit with no overlap owes $0 and is filtered out on the dashboard.
 */
import {
  addDays, lastDayOfMonth, overlapDays, parseDate, parseNamesCsv,
  parseRecurrenceConfig,
} from "./util.js";

export function unitPersonDays(occupancies, bill) {
  let total = 0;
  for (const o of occupancies) {
    const start = o.start_date instanceof Date ? o.start_date : parseDate(o.start_date);
    const end = o.end_date instanceof Date ? o.end_date : parseDate(o.end_date);
    const days = overlapDays(start, end, bill.start_date, bill.end_date);
    total += days * (o.tenant_count || 0);
  }
  return total;
}

/** Inclusive length of the bill's period in days. */
function billPeriodDays(bill) {
  return Math.round((bill.end_date - bill.start_date) / 86400000) + 1;
}

/**
 * A unit's full-occupancy reference person-days for a bill: the tenant count
 * it would have (its peak headcount among occupancies overlapping the bill)
 * times the full bill-period length. This is the denominator basis so that
 * partial occupancy is pro-rated rather than absorbing the whole bill.
 */
function unitReferencePersonDays(occupancies, bill) {
  let repTenants = 0;
  for (const o of occupancies) {
    const start = o.start_date instanceof Date ? o.start_date : parseDate(o.start_date);
    const end = o.end_date instanceof Date ? o.end_date : parseDate(o.end_date);
    if (overlapDays(start, end, bill.start_date, bill.end_date) > 0) {
      repTenants = Math.max(repTenants, o.tenant_count || 0);
    }
  }
  return repTenants * billPeriodDays(bill);
}

/**
 * Split a bill across its assigned units, pro-rated against the full period.
 *   bill: { amount, start_date (Date), end_date (Date),
 *           assignments: [{ unit_id, unit: { id, name } }, …] }
 *   occMap: { unit_id → [occupancy, …] }
 * Returns [{ unit_id, unit_name, person_days, amount }, …]
 */
export function splitBill(bill, occMap) {
  const pdByUnit = new Map();
  const refByUnit = new Map();
  for (const a of bill.assignments) {
    const occs = occMap[a.unit_id] || [];
    pdByUnit.set(a.unit_id, unitPersonDays(occs, bill));
    refByUnit.set(a.unit_id, unitReferencePersonDays(occs, bill));
  }
  let referenceTotal = 0;
  for (const v of refByUnit.values()) referenceTotal += v;
  const shares = [];
  for (const a of bill.assignments) {
    const pd = pdByUnit.get(a.unit_id) || 0;
    const amount = referenceTotal > 0 ? (pd / referenceTotal) * bill.amount : 0;
    shares.push({
      unit_id: a.unit_id,
      unit_name: a.unit ? a.unit.name : "",
      person_days: pd,
      amount: Math.round(amount * 100) / 100,
    });
  }
  return shares;
}

/** Case-insensitive membership of `kind` in an occupancy's covered_kinds. */
export function coversKind(occupancy, kind) {
  const k = String(kind || "").trim().toLowerCase();
  if (!k) return false;
  const kinds = Array.isArray(occupancy.covered_kinds)
    ? occupancy.covered_kinds
    : parseNamesCsv(occupancy.covered_kinds);
  return kinds.some((c) => String(c).trim().toLowerCase() === k);
}

/**
 * Fraction of a unit's person-days on `bill` contributed by occupancies whose
 * covered_kinds include the bill's kind — i.e. how much of the unit's share
 * is auto-covered (counted as paid). 1 when every overlapping occupancy
 * covers the kind, 0 when none does.
 */
export function coveredFraction(occupancies, bill) {
  let total = 0, covered = 0;
  for (const o of occupancies) {
    const start = o.start_date instanceof Date ? o.start_date : parseDate(o.start_date);
    const end = o.end_date instanceof Date ? o.end_date : parseDate(o.end_date);
    const pd = overlapDays(start, end, bill.start_date, bill.end_date) * (o.tenant_count || 0);
    total += pd;
    if (coversKind(o, bill.kind)) covered += pd;
  }
  return total > 0 ? covered / total : 0;
}

// ---------------- Recurring-bill generation ----------------

/**
 * Walk every period that should be generated for `rb` up to `today` (inclusive)
 * and return [(periodStart, periodEnd), …]. Returns periods that have not yet
 * passed; the caller is responsible for deduplicating against bills already
 * created (keyed by start_date).
 *
 * Deduplication anchor per recurrence:
 *   • daily  → single day
 *   • weekly → Mon–Sun period containing the trigger weekday
 *   • monthly → full calendar month
 *   • yearly  → full calendar month of each selected month
 */
export function recurringInstances(rb, today) {
  const start = rb.start_date instanceof Date ? rb.start_date : parseDate(rb.start_date);
  const config = Array.isArray(rb.recurrence_config)
    ? rb.recurrence_config
    : parseRecurrenceConfig(rb.recurrence_config);
  const endCap = rb.end_date
    ? (rb.end_date instanceof Date ? rb.end_date : parseDate(rb.end_date))
    : null;
  let cap = today;
  if (endCap && endCap < cap) cap = endCap;
  const out = [];
  if (!start) return out; // unparseable start_date: nothing to generate

  if (rb.recurrence === "daily") {
    let d = start;
    while (d <= cap) {
      out.push([d, d]);
      d = addDays(d, 1);
    }
    return out;
  }

  if (rb.recurrence === "weekly") {
    // weekday indices [0=Mon..6=Sun]; default Monday
    const active = config.length ? [...config].sort((a, b) => a - b) : [0];
    // Anchor at the Monday of `start`. JS getUTCDay(): Sun=0..Sat=6 → convert to Mon=0..Sun=6.
    const startMonOffset = (start.getUTCDay() + 6) % 7;
    let monday = addDays(start, -startMonOffset);
    while (monday <= cap) {
      const sunday = addDays(monday, 6);
      for (const wd of active) {
        const trigger = addDays(monday, wd);
        if (trigger >= start && trigger <= cap) {
          out.push([monday, sunday]);
          break;
        }
      }
      monday = addDays(monday, 7);
    }
    return out;
  }

  if (rb.recurrence === "monthly") {
    const active = config.length ? [...config].sort((a, b) => a - b) : [1];
    let y = start.getUTCFullYear();
    let m = start.getUTCMonth() + 1; // 1-based
    while (y < cap.getUTCFullYear() ||
           (y === cap.getUTCFullYear() && m <= cap.getUTCMonth() + 1)) {
      const last = lastDayOfMonth(y, m);
      for (const day of active) {
        const triggerDay = Math.min(day, last);
        const trigger = new Date(Date.UTC(y, m - 1, triggerDay));
        if (trigger >= start && trigger <= cap) {
          out.push([
            new Date(Date.UTC(y, m - 1, 1)),
            new Date(Date.UTC(y, m - 1, last)),
          ]);
          break;
        }
      }
      m += 1;
      if (m > 12) { y += 1; m = 1; }
    }
    return out;
  }

  if (rb.recurrence === "yearly") {
    const active = config.length ? [...config].sort((a, b) => a - b) : [1];
    for (let y = start.getUTCFullYear(); y <= cap.getUTCFullYear(); y++) {
      for (const month of active) {
        const billStart = new Date(Date.UTC(y, month - 1, 1));
        if (billStart >= start && billStart <= cap) {
          out.push([
            billStart,
            new Date(Date.UTC(y, month - 1, lastDayOfMonth(y, month))),
          ]);
        }
      }
    }
    return out;
  }

  return out;
}
